rdpy/testlib/test_manager.py

```diff
@@ -0,0 +1,61 @@
+from .test_client import RDPTestClientFactory
+import rdpy.core.log as log
+from threading import Event
+
+from twisted.internet import reactor, threads
+
+
+def connect(host, port, username, password, connect_timeout=10.0):
+    """
+    @summary: Connect to a RDP server with the provided details, waits until as
+    session has been established, or times out.
+    @param host: server
+    @param port: port
+    @param username:
+    @param password:
+    @param connect_timeout: Time to wait for the connection.
+    @return: RDPClientTestFactory which holds the connection to the
+    server. use factory.loggedIn to determine if it is connected.
+    """
+    log.info("Connecting to %s@%s:%d" % (username, host, port))
+    started_cond = Event()
+    factory = RDPTestClientFactory(
+        username, password,
+        started_cond)
+    reactor.callFromThread(reactor.connectTCP, host, port, factory)
```

Comment:
Is the client not running in the same thread as the twisted reactor? If not, why not? Are you trying to make it feel synchronous to the end user? If so, you should check out [`crochet`](https://github.com/itamarst/crochet) for wrapping up pure twisted code. Intermingling the reactor and threads like this has always proved... problematic for me - especially around shutdown or error conditions.